### ISSUE TITLE
18+ session option

### ISSUE THREAD

**Issue by gentlecolts** [Yellow-Dog-Man/Resonite-Issues#104]
### Is your feature request related to a problem? Please describe.

I feel like this one is rather self-explanatory, I cant even begin to count how much this has been requested

### Describe the solution you'd like

Resonite collects date of birth during account creation.  Sessions should, as an additional checkbox independent of other visibility options, provide a "18+ option", which will:

- prevent under-18 users from seeing or joining the session in any context (as if it were private for them, **even if they know or got an invite from the host**)
- prevent session thumbnail from showing in any context
- retain the session's normal visibility and joinability for users who are 18 or over

### Describe alternatives you've considered

Custom session authenticators could implement this if DoB is provided via some api, however this may present privacy implications that are more cumbersome than having it be a first-party option

### Additional Context

zangooseoo

Edit for clarity: I want it plainly understood that 18+ as a world state does not necessarily mean that NSFW content will be present, even if it means that there may be

**Comment by gentlecolts** [comment]
Some additional edge cases to consider:

- If a session was previously not set as 18+ and becomes 18+, it should kick under-18 users
- It may not be advisable to allow sessions to remove 18+ once it has been set, as that puts additional burden on users to clean up 18+ content
- To the previous point, if we were able to tag assets as 18+, they could self-delete in non-18+ sessions, much like how avatar protection self-deletes if the owning user is absent

**Comment by MargoSizemore** [comment]
If this were implemented, I believe that I should have the option to hide NSFW sessions even though I am above 18. This would be useful when using Resonite on university computers.

**Comment by FlameSoulis** [comment]
I think the most ideal situation is an option to show the sessions or hide them. In essence, an additional 'hidden' mode that is only hidden based on context.

Want to see 18+ sessions? Enable the checkbox. Want to hide them? Keep it disabled.

For education systems, just have the option hidden and set to false.

**Comment by gentlecolts** [comment]
Extra launch parameters are unnecessary and most users would not be aware of them, this is undesirable.  Having a dedicated option for 18+ specifically is important for legal compliance, so I don't believe a more generic option is suitable

**Comment by MargoSizemore** [comment]
> Extra launch parameters are unnecessary and most users would not be aware of them, this is undesirable. Having a dedicated option for 18+ specifically is important for legal compliance, so I don't believe a more generic option is suitable

But I do agree that it should be something that has to be done per computer instead of per account.

**Comment by gentlecolts** [comment]
> But I do agree that it should be something that has to be done per computer instead of per account.
I disagree, I do not believe this needs to be more complex than an additional checkbox on the session tab.  Please elaborate on why you feel that way

**Comment by MargoSizemore** [comment]
> > But I do agree that it should be something that has to be done per computer instead of per account.
> 
> I disagree, I do not believe this needs to be more complex than an additional checkbox on the session tab. Please elaborate on why you feel that way

I feel this way because Resonite and Neos are on some of my university's computers. It's not like these computers are specifically for the BSDM club or anything, so I think it's understandable for me to want to avoid NSFW content on these university computers, wouldn't you agree?

**Comment by gentlecolts** [comment]
> I feel this way because Resonite and Neos are on some of my university's computers. It's not like these computers are specifically for the BSDM club or anything, so I think it's understandable for me to want to avoid NSFW content on these university computers, wouldn't you agree?

I think we're talking about two different features.  Sessions having support for a proper "18+" flag is a separate issue than wanting to opt in/out of seeing those sessions presuming you are in fact 18+.  Also, 18+ does not necessarily mean NSFW.  I regularly host SFW 18+ gatherings

**Comment by MargoSizemore** [comment]
> > I feel this way because Resonite and Neos are on some of my university's computers. It's not like these computers are specifically for the BSDM club or anything, so I think it's understandable for me to want to avoid NSFW content on these university computers, wouldn't you agree?
> 
> I think we're talking about two different features. Sessions having support for a proper "18+" flag is a separate issue than wanting to opt in/out of seeing those sessions presuming you are in fact 18+. Also, 18+ does not necessarily mean NSFW. I regularly host SFW 18+ gatherings

Okay. Then in that case, NSFW sessions should require a very purposeful opt in per computer. 18+ sessions should always be visible to people above 18. I hope this fully articulates exactly what I'm thinking.

**Comment by The-furf-of-July** [comment]
> I feel this way because Resonite and Neos are on some of my university's computers. It's not like these computers are specifically for the BSDM club or anything, so I think it's understandable for me to want to avoid NSFW content on these university computers, wouldn't you agree?

They have a point about the opt-in per computer thing, in the case of shared or public devices. Perhaps the setting should be saved per account AND device ID? That way if a user has the setting enabled on their home computer, then goes to a different computer they need to enable it again for that device, without changing their setting at home.

> Okay. Then in that case, NSFW sessions should require a very purposeful opt in per computer. 18+ sessions should always be visible to people above 18. I hope this fully articulates exactly what I'm thinking.

I do not think separating 18+ and NSFW as settings is a good idea, for ease of implementation and use. All ideas here so far still fit as a generic "18+", i.e. you wouldn't put an 18+ tag on an item unless you wanted to ensure minors would be unable to view or download it, whatever that reason may be.

**Comment by shiftyscales** [comment]
As Frooxius mentioned in #1022:

> You can already mark sessions as "Hidden", which hides them from the view and will prevent previews from being shown to users.
> 
> Marking it explicitly as NSFW has larger implications and probably not something we will add in the immediate future. There's a number of issues to be resolved with that:
> - We need to figure out the legalities of this, because this might put some responsibility on our end (such as age verification)
> - Policy on what's NSFW and what isn't - the line can be difficult to draw for some things and we'll need to moderate the sessions in some way, so this might not even resolve the issue that is currently being tackled - if what you consider uncomfortable is not considered as NSFW, this option will not help and is not the right solution for this
> - What would this actually do on top of hiding the session (which you can already do)? Would it prevent underage users from joining? What standard would we need to use to ensure of that?
> 
> There is a lot of issues that we need to work through if we decide we'd want such option to happen. 
> 
> I'd just recommend using "Hidden" session for now if you don't want it to be displayed.

There are a lot of considerations and implications that would need to be discussed and decided on before we can consider adding such functionality to the platform.

For now, I am going to close and merge this issue into the issue I linked above as that is where more recent discussion on this topic has occurred- notably from other team members.
